fix(DAV): Clean-up remote + handle application/xml properly

joshtrichards opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

## Summary

- fix: Handle `Content-Type` `application/xml` (in remote in addition to `text/xml`)

Also: 

- fix: Handle if service doesn't resolve properly
- fix: Return a 503 (rather than 200) if app is not enabled/installed [note: public.php should probably get this fix too]

Other than that, some general refactoring.

To ease review and for general comparison, compare to `public.php`.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

joshtrichards committed:
fix(DAV): Handle application/xml in remote properly

Also: 
- fix: Handle if service doesn't resolve properly
- fix: Return a 503 (rather than 200) if app is not enabled/installed

Other than that, some general refactoring.

To ease review and for general comparison, compare to `public.php`.

Signed-off-by: Josh <josh.t.richards@gmail.com>

kesselb commented on remote.php:
```diff
@@ -29,23 +31,23 @@ class RemoteException extends \Exception {
+		$request = Server::get(IRequest::class);
+		// treat XML as a WebDAV request
+		$contentType = (string)$request->getHeader('Content-Type');
+		if (str_contains($contentType, 'application/xml') || str_contains($contentType, 'text/xml')) { // TODO: consider centralizing this repetitive check
```
As we have that check in a couple of places, a small helper would indeed be nice.

kesselb reviewed (commented):
Thank you

joshtrichards commented on remote.php:
```diff
@@ -29,23 +31,23 @@ class RemoteException extends \Exception {
-		if ($isXmlContentType === 0) {
+		$request = Server::get(IRequest::class);
+		// treat XML as a WebDAV request
+		$contentType = (string)$request->getHeader('Content-Type');
```
I agree, but for some reason we're doing that everywhere else we do a similar check so I did it for consistency.

kesselb commented on remote.php:
```diff
@@ -99,54 +101,55 @@ function resolveService($service) {
-				throw new RemoteException('App not installed: ' . $app);
-			}
-			$appManager->loadApp($app);
-			$file = $appManager->getAppPath($app) . '/' . ($parts[1] ?? '');
```
Thanks, sounds sane :+1:

joshtrichards committed:
chore: fixup remote.php

Co-authored-by: Daniel <mail@danielkesselberg.de>
Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: re-organize remote.php

Whatever is settled on here will likely get ported to /public.php as well.

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: drop unused remote_ dynamic resolver

Endpoints are all hardcoded since https://github.com/nextcloud/server/commit/7bdde98617ad3c7f56dced1add0a155f4a0e5a2e and were formally deprecated within info.xml in nextcloud/appstore#156. 

Follow-up PR will clean-up references elsewhere.

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
fix: fixup for lint

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: Update psalm-baseline.xml

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: unify public.php with updated remote.php

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
chore: drop unused IConfig from remote.php

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
Merge branch 'master' into jtr-refactor-remote-php

Signed-off-by: Josh <josh.t.richards@gmail.com>
